types/module/module.go

```diff
@@ -383,27 +383,34 @@ func (m *Manager) ExportGenesis(ctx sdk.Context, cdc codec.JSONCodec) map[string
 // ExportGenesisForModules performs export genesis functionality for modules
 func (m *Manager) ExportGenesisForModules(ctx sdk.Context, cdc codec.JSONCodec, modulesToExport []string) map[string]json.RawMessage {
 	genesisData := make(map[string]json.RawMessage)
-	if len(modulesToExport) == 0 {
-		for _, moduleName := range m.OrderExportGenesis {
-			if module, ok := m.Modules[moduleName].(HasGenesis); ok {
-				genesisData[moduleName] = module.ExportGenesis(ctx, cdc)
-			}
-		}
 
-		return genesisData
+	if len(modulesToExport) == 0 {
+		modulesToExport = m.OrderExportGenesis
 	}
 
 	// verify modules exists in app, so that we don't panic in the middle of an export
 	if err := m.checkModulesExists(modulesToExport); err != nil {
 		panic(err)
 	}
 
-	for _, moduleName := range modulesToExport {
+	channels := make([]chan json.RawMessage, len(modulesToExport))
+	modulesWithGenesis := make([]string, 0, len(modulesToExport))
+
+	for i, moduleName := range modulesToExport {
 		if module, ok := m.Modules[moduleName].(HasGenesis); ok {
-			genesisData[moduleName] = module.ExportGenesis(ctx, cdc)
+			channels[i] = make(chan json.RawMessage)
+			modulesWithGenesis = append(modulesWithGenesis, moduleName)
+
+			go func(module HasGenesis, ch chan json.RawMessage) {
+				ch <- module.ExportGenesis(ctx, cdc)
+			}(module, channels[i])
 		}
 	}
 
+	for i, moduleName := range modulesWithGenesis {
+		genesisData[moduleName] = <-channels[i]
```

Comment:
You are iterating over `modulesWithGenesis` now which may be a subset of `modulesToExport`. The `i` value therefore is not the same. 
For example if we have `[foo`, `bar`] as modules in `modulesToExport` with `foo` not implementing `HasGenesis` then `modulesWithGenesis` has only the `bar` element and you would still wait for `channel[0]` to return. 
Some tests on this feature would be very useful